qa/rpc-tests/wallet_protectcoinbase.py

```diff
@@ -0,0 +1,110 @@
+                sleep(1)
+            else:
+                status = results[0]["status"]
+                break
+        print('...returned status: {}'.format(status))
+        assert_equal("success", status)
+
+    def run_test (self):
+        print "Mining blocks..."
+
+        self.nodes[0].generate(4)
+
+        walletinfo = self.nodes[0].getwalletinfo()
+        assert_equal(walletinfo['immature_balance'], 40)
+        assert_equal(walletinfo['balance'], 0)
+
+        self.sync_all()
+        self.nodes[1].generate(101)
+        self.sync_all()
+
+        assert_equal(self.nodes[0].getbalance(), 40)
+        assert_equal(self.nodes[1].getbalance(), 10)
+        assert_equal(self.nodes[2].getbalance(), 0)
+
+        # Send will fail because we are enforcing the consensus rule that
+        # coinbase utxos can only be sent to a zaddr.
+        errorString = ""
+        try:
+            self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 1.0)
+        except JSONRPCException,e:
+            errorString = e.error['message']
+        assert_equal("Coinbase funds can only be sent to a zaddr" in errorString, True)
+
+        # send node 0 taddr to node 0 zaddr
+        mytaddr = self.nodes[0].getnewaddress()
+        myzaddr = self.nodes[0].z_getnewaddress()
+        recipients = []
+        recipients.append({"address":myzaddr, "amount":20.0})
+        myopid = self.nodes[0].z_sendmany(mytaddr, recipients)
+        self.wait_for_operationd_success(myopid)
+        self.nodes[1].generate(1)
+        self.sync_all()
+
+        # convert note to transparent funds
+        recipients = []
+        recipients.append({"address":mytaddr, "amount":20.0})
+        myopid = self.nodes[0].z_sendmany(myzaddr, recipients)
+        self.wait_for_operationd_success(myopid)
+        self.nodes[1].generate(1)
+        self.sync_all()
+
+        # check balances
+        resp = self.nodes[0].z_gettotalbalance()
+        assert_equal(Decimal(resp["transparent"]), Decimal('30.0'))
+        assert_equal(Decimal(resp["private"]), Decimal('9.9998'))
+        assert_equal(Decimal(resp["total"]), Decimal('39.9998'))
+
+        # Send will fail because send amount is too big, even when including coinbase utxos
+        errorString = ""
+        try:
+            self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 99999.0)
+        except JSONRPCException,e:
+            errorString = e.error['message']
+        print "error: ", errorString
+        assert_equal("Insufficient funds" in errorString, True)
+
+        # Send will fail because of insufficient funds unless sender uses coinbase utxos
+        try:
+            self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), 21.0)
+        except JSONRPCException,e:
+            errorString = e.error['message']
+        print "error: ", errorString
+        assert_equal("Insufficient funds, coinbase funds can only be spent after they have been sent to a zaddr" in errorString, True)
```

Comment:
Does this test need another case in which the send succeeds because there are enough non-coinbase funds, but the coin selection prior to this PR would have chosen a coinbase input?

(Sorry to raise this after already having ACKed; I know we're short on time.)